### PR TITLE
Add faster TCP checksum implementation for x86_64

### PR DESCRIPTION
**Pull request by BurntPizza**
I was inspired/nerdsniped by your comment in the reddit thread on the TCP checksum being a bottleneck.
With this I'm getting a ~8-10% speed increase on the benchmark example.

It took some fiddling to get LLVM to vectorize satisfactorily, and it would still be possible to improve on the codegen using actual SIMD intrinsics. You can see the result [here](https://godbolt.org/g/9nZsd1). With `sse4.1` or `avx` target features the inner loop gets much tighter still.
One thing is that I only have a `x86_64` machine, so that's why it's the only arch enabled. I'm curious how it would do on other architectures that have vector instructions.

Please evaluate and benchmark (you can use my repo) and feel free to decline/request changes if it's not as good as I think it is.

### CODE REVIEW

**Comment by kpp** [comment]
Why not use https://github.com/AdamNiederer/faster ? @AdamNiederer seems like a good challenge for you!

**Comment by whitequark** [comment]
Sweet! Can you please add some testcases?

@kpp that doesn't work on `#![no_std]` unfortunately, which is smoltcp's primary deployment target.

**Comment by AdamNiederer** [comment]
I don't think there's anything *preventing* it working with `no_std` - I'll see if I can hack something together.

**Comment by BurntPizza** [comment]
I've revised to hopefully improve further upon  275eb90785a26ed1bd942f087c2d81647dc58f84, which was apparently the one simple formulation which works that I didn't try last night. :\

You'll want to bench this before accepting the PR, as it's pretty close.

What kind of testcases are you looking for?

**Comment by whitequark** [comment]
> What kind of testcases are you looking for?

Not anymore, since the correctness is now pretty much obvious.

File: src/wire/ip.rs [hunk 1]
```diff
@@ -611,15 +611,29 @@ pub mod checksum {
 
     /// Compute an RFC 1071 compliant checksum (without the final complement).
     pub fn data(mut data: &[u8]) -> u16 {
-        // See RFC 1071 section 4.1 for the original implementation.
-        let mut accum: u32 = 0;
+        let mut accum = 0;
+
+        // Operate on large chunks first to ease the job of the autovectorizer.
```
Comment (kpp) [inline src/wire/ip.rs]: I would like to see comments like:

```
each 32-byte chunk
  take by 2 bytes as network order u16
  sum them
sum the rest that does not fit the last 32-byte chunk
```

Comment (whitequark) [reply src/wire/ip.rs]: Better?

**Comment by 0x7CFE** [comment]
> that doesn't work on #![no_std] unfortunately, which is smoltcp's primary deployment target.

@whitequark, now it does! https://adamniederer.com/blog/faster-pr-2.html

**Comment by AdamNiederer** [comment]
I'm seeing a 2x speedup on AVX with `faster`, and a 20-40% speedup on SSE using an unreleased verison of `faster`. I'll send a PR once 0.5.0 is out.

**Pull request merged**